tests/model_fields/test_jsonfield.py

```diff
@@ -0,0 +1,657 @@
+        obj = NullableJSONModel.objects.create(value={'j': 1})
+        if connection.vendor == 'oracle':
+            # On Oracle, the query returns JSON objects and arrays that do not have a 'null' value
+            # at the specified path, including those that do not have the key.
+            self.assertSequenceEqual(
+                NullableJSONModel.objects.exclude(value__j=None),
+                self.object_data[:3] + self.object_data[4:] + [obj]
+            )
+        else:
+            self.assertSequenceEqual(NullableJSONModel.objects.exclude(value__j=None), [obj])
+
+    def test_isnull_key_or_none(self):
+        obj = NullableJSONModel.objects.create(value={'a': None})
+        self.assertSequenceEqual(
+            NullableJSONModel.objects.filter(Q(value__a__isnull=True) | Q(value__a=None)),
+            self.scalar_data + self.object_data[:2] + self.object_data[4:] + [obj]
+        )
+
+    def test_shallow_list_lookup(self):
+        self.assertSequenceEqual(
+            NullableJSONModel.objects.filter(value__0=1),
+            [self.object_data[4]]
+        )
+
+    def test_shallow_obj_lookup(self):
+        self.assertSequenceEqual(
+            NullableJSONModel.objects.filter(value__a='b'),
+            [self.object_data[2], self.object_data[3]]
+        )
+
+    def test_deep_lookup_objs(self):
+        self.assertSequenceEqual(
+            NullableJSONModel.objects.filter(value__k__l='m'),
+            [self.object_data[3]]
+        )
+
+    def test_shallow_lookup_obj_target(self):
+        self.assertSequenceEqual(
+            NullableJSONModel.objects.filter(value__k={'l': 'm'}),
+            [self.object_data[3]]
+        )
+
+    def test_deep_lookup_array(self):
+        self.assertSequenceEqual(
+            NullableJSONModel.objects.filter(value__1__0=2),
+            [self.object_data[4]]
+        )
+
+    def test_deep_lookup_mixed(self):
+        self.assertSequenceEqual(
+            NullableJSONModel.objects.filter(value__d__1__f='g'),
+            [self.object_data[3]]
+        )
+
+    def test_deep_lookup_transform(self):
+        self.assertSequenceEqual(
+            NullableJSONModel.objects.filter(value__c__gt=2),
+            [self.object_data[2], self.object_data[3]]
+        )
+        self.assertSequenceEqual(
+            NullableJSONModel.objects.filter(value__c__lt=5),
+            []
+        )
+
+    def test_usage_in_subquery(self):
+        self.assertSequenceEqual(
+            NullableJSONModel.objects.filter(id__in=NullableJSONModel.objects.filter(value__c=14)),
+            self.object_data[2:4]
+        )
+
+    def test_iexact(self):
+        self.assertIs(NullableJSONModel.objects.filter(value__foo__iexact='BaR').exists(), True)
+        self.assertIs(NullableJSONModel.objects.filter(value__foo__iexact='"BaR"').exists(), False)
+
+    def test_icontains(self):
+        self.assertIs(NullableJSONModel.objects.filter(value__foo__icontains='"bar"').exists(), False)
```

Comment:
`__icontains` is not properly tested it should check:
```python
NullableJSONModel.objects.filter(value__foo__icontains='A').exists()
```
also tests for `KeyTransformContains` are missing.